src/NKSolver/NKSolvers.F90

```diff
@@ -2162,27 +2028,293 @@ subroutine FormJacobianANK
+                        do i = 2, il
 
-            ! Extension for setting coarse grids:
-            if (useCoarseMats) then
-                do lvl = 2, agmgLevels
-                    call MatSetValuesBlocked(A(lvl), 1, coarseRows(lvl), 1, coarseRows(lvl), &
-                                             blk, ADD_VALUES, ierr)
+                            ! Compute the block for this cell
+                            call computeTimeStepBlock(i, j, k, timeStepBlock)
+
+                            ! Get the global cell index
+                            irow = globalCell(i, j, k)
+
+                            ! Add the contribution to the PETSc matrix
+                            call MatSetValuesBlocked(timeStepInv, 1, irow, 1, irow, timeStepBlock, ADD_VALUES, ierr)
+                            call EChk(ierr, __FILE__, __LINE__)
+
+                            ! Extension for setting coarse grids
+                            if (useCoarseMats) then
+                                do lvl = 2, agmgLevels
+                                    coarseRows(lvl) = coarseIndices(nn, lvl - 1)%arr(i, j, k)
+                                    call MatSetValuesBlocked(A(lvl), 1, coarseRows(lvl), 1, coarseRows(lvl), &
+                                                             timeStepBlock, ADD_VALUES, ierr)
+                                    call EChk(ierr, __FILE__, __LINE__)
+                                end do
+                            end if
+
+                        end do
+                    end do
                 end do
+            end do
+        end do
+
+        ! PETSc Matrix Assembly
+        call MatAssemblyBegin(timeStepInv, MAT_FINAL_ASSEMBLY, ierr)
+        call EChk(ierr, __FILE__, __LINE__)
+        call MatAssemblyEnd(timeStepInv, MAT_FINAL_ASSEMBLY, ierr)
+        call EChk(ierr, __FILE__, __LINE__)
+
+    end subroutine computeTimeStepInv
+
+    subroutine computeTimeStepBlock(i, j, k, timeStepBlock)
+        ! Computes the inverse time step block matrix for a given cell i, j, k.
+
+        use constants
+        use inputPhysics, only: machInf => mach
+        use blockPointers, only: volRef, w, dtl, gamma, p, aa
+        use flowVarRefState, only: viscous, nt1
+        use inputIteration, only: turbResScale
+        use communication
+        implicit none
+
+        ! Input variables
+        integer(kind=intType), intent(in) :: i, j, k
+
+        ! Output variables
+        real(kind=realType), dimension(nState, nState), intent(out) :: timeStepBlock
+
+        ! Local variables
+        real(kind=realType) :: blendFactor, dtInv, rho, velX, velY, velZ
+        real(kind=realType) :: speed, speedOfSound, mach, machSqr, machSqrTrunc, alpha, beta, tau, gammaMinusOne
+        real(kind=realType) :: speedXY, sinTheta, cosTheta, sinAlpha, cosAlpha
+        real(kind=realType), dimension(nState, nState) :: streamToCart, symmToCons, consToSymm, stateToCons
+
+        ! Zero the block matrices
```

Comment:
By any chance, will these block matrices run out of memory?

Reply:
The block matrices are 5x5 or 6x6, so they don't use much memory. Also, these block matrices are only allocated for one cell at a time.